File: Tutorial4-4ComponentBuildersMultipleModules/src/main/java/com/example/tutorial4_4componentbuildersmultiplemodules/di/AppComponent.java
package com.example.tutorial4_4componentbuildersmultiplemodules.di;


import com.example.tutorial4_4componentbuildersmultiplemodules.MainActivity;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {AppModule.class, ToastMakerModule.class, SensorControllerModule.class})
public interface AppComponent {
    void inject(MainActivity mainActivity);

    // DaggerAppComponent.build() returns this Builder interface

    @Component.Builder
    interface Builder {
        AppComponent build();

        Builder appModule(AppModule appModule);

        Builder sensorControllerModule(SensorControllerModule sensorControllerModule);

        Builder toastMakerModule(ToastMakerModule toastMakerModule);
    }

}

File: Tutorial4-4ComponentBuildersMultipleModules/src/main/java/com/example/tutorial4_4componentbuildersmultiplemodules/di/AppComponent.kt
package com.example.tutorial4_4componentbuildersmultiplemodules.di

import com.example.tutorial4_4componentbuildersmultiplemodules.MainActivity
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [AppModule::class, ToastMakerModule::class, SensorControllerModule::class])
interface AppComponent {

    fun inject(mainActivity: MainActivity)

    // DaggerAppComponent.build() returns this Builder interface
    @Component.Builder
    interface Builder {
        fun build(): AppComponent
        fun appModule(appModule: AppModule): Builder
        fun sensorControllerModule(sensorControllerModule: SensorControllerModule): Builder
        fun toastMakerModule(toastMakerModule: ToastMakerModule): Builder
    }
}